pitop/processing/algorithms/faces/face_detector.py

```diff
@@ -0,0 +1,245 @@
+            if face_center is None:
+                self.__stop_tracker()
+                # attempt to detect face since tracker has failed
+                face_rectangle, face_center, face_features = self.__detect_largest_face(frame=frame_to_process)
+
+        self.face = self.__prepare_face_data(frame=frame,
+                                             face=self.face,
+                                             rectangle=face_rectangle,
+                                             center=face_center,
+                                             features=face_features)
+        if self._print_fps:
+            self._fps.update()
+
+        return self.face
+
+    def __get_frame_to_process(self, frame):
+        """Resize frame, convert to grayscale and use contrast limited adaptive
+        histogram equalization (CLAHE) to improve the contrast of the image so
+        that face features are more pronounced.
+
+        :param frame: original frame from the camera in OpenCV format
+        :return: OpenCV image to send to face processing algorithms
+        """
+        return self._clahe_filter.apply(
+            cv2.cvtColor(
+                resize(frame, width=self._image_processing_width),
+                cv2.COLOR_BGR2GRAY
+            )
+        )
+
+    def __detect_largest_face(self, frame):
+        """Use dlib face rectangle detector on the frame and process the
+        detected rectangles to get the face data.
+
+        :param frame: Image frame to process for faces
+        :return: Detected face data
+        """
+        face_rectangle, face_center, face_features = self.__process_detected_rectangles(
+            frame,
+            self._face_rectangle_detector(frame, self._FACE_DETECTOR_PYRAMID_LAYERS)
+        )
+
+        return face_rectangle, face_center, face_features
+
+    def __process_detected_rectangles(self, frame, rectangles_dlib):
+        """Find largest face rectangle and process it to get the required face
+        data.
+
+        :param frame: OpenCV frame to use for processing
+        :param rectangles_dlib: Rectangles found using dlib's face detector in the dlib Rectangle format.
+        :return: Detected face data
+        """
+        if len(rectangles_dlib) == 0:
+            return None, None, None
+
+        largest_rectangle_dlib = None
+        area = 0
+        face_rectangle = None
+        face_center = None
+        for rectangle_dlib in rectangles_dlib:
+            x, y, w, h = face_utils.rect_to_bb(rectangle_dlib)
+            current_area = w * h
+            if current_area > area:
+                area = current_area
+                largest_rectangle_dlib = rectangle_dlib
+                face_rectangle = (x, y, w, h)
+                face_center = (int(x + w / 2), int(y + h / 2))
+
+        if rectangles_dlib is None:
+            # this should never trigger. Discuss.
+            return None, None, None
```

Comment:
if this should never trigger, why not just raise an exception? something must be wrong somewhere, right?

Reply:
I've removed it completely. It's basically impossible for it to trigger unless the width and height of the returned rectangle is 0, which I think we can safely assume will never be the case

Reply:
Also FYI this was supposed to be checking `largest_rectangle_dlib`, not `rectangles_dlib` - that was a typo anyway and was even less likely since dlib face detector always returns a list